File: android/src/main/java/me/lazerka/mf/android/activity/EventLogger.java
/*
 *     Find Us: privacy oriented location tracker for your friends and family.
 *     Copyright (C) 2016 Dzmitry Lazerka dlazerka@gmail.com
 *
 *     This program is free software; you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation; either version 2 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License along
 *     with this program; if not, write to the Free Software Foundation, Inc.,
 *     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

package me.lazerka.mf.android.activity;

import android.os.Bundle;
import com.google.firebase.analytics.FirebaseAnalytics;

import javax.annotation.Nullable;

public class EventLogger {
	private final String eventName;
	@Nullable // in error case
	private final FirebaseAnalytics firebaseAnalytics;
	private final Bundle bundle = new Bundle();

	public EventLogger(String eventName, @Nullable FirebaseAnalytics firebaseAnalytics) {
		this.eventName = eventName;
		this.firebaseAnalytics = firebaseAnalytics;
	}

	public EventLogger param(String key, String value) {
		bundle.putString(key, value);
		return this;
	}

	public EventLogger param(String key, int value) {
		bundle.putInt(key, value);
		return this;
	}

	public EventLogger param(String key, long value) {
		bundle.putLong(key, value);
		return this;
	}

	public EventLogger param(String key, boolean value) {
		bundle.putBoolean(key, value);
		return this;
	}

	public void send() {
		if (firebaseAnalytics != null) {
			firebaseAnalytics.logEvent(eventName, bundle);
		}
	}
}

File: android/src/main/kotlin/com/baraded/mf/logging/EventLogger.kt
/*
 *     Copyright (C) 2017 Dzmitry Lazerka
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

package com.baraded.mf.logging;

import android.os.Bundle
import com.google.firebase.analytics.FirebaseAnalytics
import java.util.regex.Pattern


class EventLogger(
        private val eventName: String,
        private val firebaseAnalytics: FirebaseAnalytics
) {
    private val bundle = Bundle()

    init {
        checkName(eventName)
    }

    fun param(key: String, value: String): EventLogger {
        checkKey(key)
        bundle.putString(key, value)
        return this
    }

    fun param(key: String, value: Int): EventLogger {
        checkKey(key)
        bundle.putInt(key, value)
        return this
    }

    fun param(key: String, value: Long): EventLogger {
        checkKey(key)
        bundle.putLong(key, value)
        return this
    }

    fun param(key: String, value: Boolean): EventLogger {
        checkKey(key)
        bundle.putBoolean(key, value)
        return this
    }

    fun send() {
        // It logs the event to Logcat as well.
        firebaseAnalytics.logEvent(eventName, bundle)
    }

    companion object {
        /**
         * Otherwise FirebaseCrash errors "Name must consist of letters, digits or _ (underscores)."
         * for both event name and param name.
         */
        private val NAME = Pattern.compile("^[a-z0-9_]{1,32}$", Pattern.CASE_INSENSITIVE)
    }

    private fun checkKey(key: String) {
        if (!NAME.matcher(key).matches()) {
            throw IllegalArgumentException("Event key " + key)
        }
    }

    private fun checkName(name: String) {
        if (!NAME.matcher(name).matches()) {
            throw IllegalArgumentException("Event name: " + name)
        }
    }
}